File: find_duplicates.cs
// find_duplicates.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using System.Text.RegularExpressions;

class FindDuplicates
{
    static string Colorize(string text, string color)
    {
        string col = color switch
        {
            "green" => "\x1b[92m",
            "red" => "\x1b[91m",
            "yellow" => "\x1b[93m",
            "blue" => "\x1b[94m",
            _ => "\x1b[0m"
        };
        return col + text + "\x1b[0m";
    }

    static string GetFileHash(string path, string algo, long partial)
    {
        using var fs = new FileStream(path, FileMode.Open, FileAccess.Read);
        HashAlgorithm hash = algo.ToLower() switch
        {
            "md5" => MD5.Create(),
            "sha1" => SHA1.Create(),
            _ => throw new NotSupportedException()
        };
        byte[] buffer = new byte[8192];
        long total = 0;
        int read;
        while ((read = fs.Read(buffer, 0, buffer.Length)) > 0 && (partial == 0 || total < partial))
        {
            if (partial > 0 && total + read > partial)
                read = (int)(partial - total);
            hash.TransformBlock(buffer, 0, read, buffer, 0);
            total += read;
            if (partial > 0 && total >= partial) break;
        }
        hash.TransformFinalBlock(buffer, 0, 0);
        return BitConverter.ToString(hash.Hash).Replace("-", "").ToLower();
    }

    static void FindDuplicates(string root, bool recursive, string algo, long partial,
                               List<string> excludePatterns,
                               Dictionary<string, List<List<string>>> duplicates)
    {
        var sizeMap = new Dictionary<long, List<string>>();
        void Walk(string dir)
        {
            foreach (var entry in Directory.EnumerateFileSystemEntries(dir))
            {
                if (Directory.Exists(entry))
                {
                    if (!recursive && entry != root) continue;
                    bool skip = false;
                    foreach (var pat in excludePatterns)
                        if (Regex.IsMatch(Path.GetFileName(entry), WildcardToRegex(pat))) { skip = true; break; }
                    if (skip) continue;
                    Walk(entry);
                }
                else
                {
                    bool skip = false;
                    foreach (var pat in excludePatterns)
                        if (Regex.IsMatch(Path.GetFileName(entry), WildcardToRegex(pat))) { skip = true; break; }
                    if (skip) continue;
                    var info = new FileInfo(entry);
                    if (info.Length > 0)
                    {
                        if (!sizeMap.ContainsKey(info.Length))
                            sizeMap[info.Length] = new List<string>();
                        sizeMap[info.Length].Add(entry);
                    }
                }
            }
        }
        Walk(root);

        foreach (var kv in sizeMap)
        {
            if (kv.Value.Count < 2) continue;
            var hashMap = new Dictionary<string, List<string>>();
            foreach (var f in kv.Value)
            {
                var h = GetFileHash(f, algo, partial);
                if (!hashMap.ContainsKey(h))
                    hashMap[h] = new List<string>();
                hashMap[h].Add(f);
            }
            foreach (var hkv in hashMap)
            {
                if (hkv.Value.Count > 1)
                {
                    if (!duplicates.ContainsKey(hkv.Key))
                        duplicates[hkv.Key] = new List<List<string>>();
                    duplicates[hkv.Key].Add(hkv.Value);
                }
            }
        }
    }

    static string WildcardToRegex(string pattern) =>
        "^" + Regex.Escape(pattern).Replace("\\*", ".*").Replace("\\?", ".") + "$";

    static string FormatSize(long size)
    {
        string[] units = { "B", "KB", "MB", "GB", "TB" };
        double s = size;
        int i = 0;
        while (s >= 1024 && i < units.Length - 1) { s /= 1024; i++; }
        return $"{s:F1} {units[i]}";
    }

    static void PrintDuplicates(Dictionary<string, List<List<string>>> duplicates, bool verbose)
    {
        if (duplicates.Count == 0)
        {
            Console.WriteLine(Colorize("Дубликатов не найдено.", "green"));
            return;
        }
        int totalGroups = 0, totalFiles = 0;
        long totalSize = 0;
        foreach (var kv in duplicates)
        {
            foreach (var group in kv.Value)
            {
                totalGroups++;
                totalFiles += group.Count;
                var size = new FileInfo(group[0]).Length;
                totalSize += size * (group.Count - 1);
                Console.WriteLine(Colorize($"\nГруппа дубликатов (размер: {FormatSize(size)}):", "blue"));
                Console.WriteLine($"  Хеш: {kv.Key}");
                for (int i = 0; i < group.Count; i++)
                {
                    if (i == 0)
                        Console.WriteLine(Colorize($"  [оригинал] {group[i]}", "green"));
                    else
                        Console.WriteLine(Colorize($"  [дубликат] {group[i]}", "yellow"));
                }
                if (verbose) Console.WriteLine($"  Количество файлов: {group.Count}");
            }
        }
        Console.WriteLine(Colorize($"\nВсего групп: {totalGroups}, файлов: {totalFiles}, экономия: {FormatSize(totalSize)}", "green"));
    }

    static void Main(string[] args)
    {
        string root = ".";
        bool recursive = true, deleteDups = false, hardlink = false, dryRun = false, verbose = false, yes = false;
        string algo = "md5", moveTo = null, logFile = null;
        long partial = 0;
        var excludePatterns = new List<string>();

        for (int i = 0; i < args.Length; i++)
        {
            string arg = args[i];
            if (arg == "-p" && i+1 < args.Length) root = args[++i];
            else if (arg == "-r") recursive = true;
            else if (arg == "-a" && i+1 < args.Length) algo = args[++i];
            else if (arg == "--partial" && i+1 < args.Length) partial = long.Parse(args[++i]);
            else if (arg == "-e" && i+1 < args.Length) excludePatterns.AddRange(args[++i].Split(','));
            else if (arg == "--delete") deleteDups = true;
            else if (arg == "--move" && i+1 < args.Length) moveTo = args[++i];
            else if (arg == "--hardlink") hardlink = true;
            else if (arg == "-n") dryRun = true;
            else if (arg == "-l" && i+1 < args.Length) logFile = args[++i];
            else if (arg == "-v") verbose = true;
            else if (arg == "-y") yes = true;
            else if (arg == "-h") { Console.WriteLine("Help..."); return; }
        }

        if (!Directory.Exists(root))
        {
            Console.WriteLine(Colorize($"Ошибка: '{root}' не является папкой", "red"));
            return;
        }

        var duplicates = new Dictionary<string, List<List<string>>>();
        FindDuplicates(root, recursive, algo, partial, excludePatterns, duplicates);
        PrintDuplicates(duplicates, verbose);

        if (deleteDups || moveTo != null || hardlink)
        {
            if (!yes)
            {
                Console.Write(Colorize("Выполнить действия над дубликатами? [y/N] ", "yellow"));
                var ans = Console.ReadLine();
                if (ans?.ToLower() != "y")
                {
                    Console.WriteLine(Colorize("Операция отменена.", "red"));
                    return;
                }
            }
            foreach (var groups in duplicates.Values)
            {
                foreach (var group in groups)
                {
                    var original = group[0];
                    for (int i = 1; i < group.Count; i++)
                    {
                        var dup = group[i];
                        if (dryRun)
                        {
                            Console.WriteLine(Colorize($"[DRY RUN] Будет обработан дубликат: {dup}", "yellow"));
                            continue;
                        }
                        if (deleteDups)
                        {
                            File.Delete(dup);
                            Console.WriteLine(Colorize($"Удалён: {dup}", "red"));
                        }
                        else if (moveTo != null)
                        {
                            Directory.CreateDirectory(moveTo);
                            var dest = Path.Combine(moveTo, Path.GetFileName(dup));
                            File.Move(dup, dest);
                            Console.WriteLine(Colorize($"Перемещён: {dup} -> {dest}", "yellow"));
                        }
                        else if (hardlink)
                        {
                            File.Delete(dup);
                            File.CreateSymbolicLink(dup, original);
                            Console.WriteLine(Colorize($"Создана жёсткая ссылка: {dup} -> {original}", "blue"));
                        }
                    }
                }
            }
        }
    }
}

File: find_duplicates.java
// find_duplicates.java
import java.io.*;
import java.nio.file.*;
import java.security.*;
import java.util.*;
import java.util.regex.Pattern;

public class find_duplicates {
    private static final String RESET = "\u001B[0m";
    private static final String GREEN = "\u001B[92m";
    private static final String RED = "\u001B[91m";
    private static final String YELLOW = "\u001B[93m";
    private static final String BLUE = "\u001B[94m";

    private static String colorize(String text, String color) {
        return color + text + RESET;
    }

    private static String getFileHash(String path, String algo, long partial) throws Exception {
        MessageDigest md = MessageDigest.getInstance(algo.toUpperCase());
        try (FileInputStream fis = new FileInputStream(path)) {
            byte[] buffer = new byte[8192];
            long total = 0;
            int read;
            while ((read = fis.read(buffer)) != -1 && (partial == 0 || total < partial)) {
                if (partial > 0 && total + read > partial)
                    read = (int)(partial - total);
                md.update(buffer, 0, read);
                total += read;
            }
        }
        byte[] digest = md.digest();
        StringBuilder sb = new StringBuilder();
        for (byte b : digest) sb.append(String.format("%02x", b));
        return sb.toString();
    }

    private static void walk(Path dir, boolean recursive, List<String> excludePatterns,
                             Map<Long, List<String>> sizeMap) throws IOException {
        try (DirectoryStream<Path> stream = Files.newDirectoryStream(dir)) {
            for (Path entry : stream) {
                if (Files.isDirectory(entry)) {
                    if (!recursive && !entry.equals(dir)) continue;
                    boolean skip = false;
                    for (String pat : excludePatterns)
                        if (Pattern.matches(pat, entry.getFileName().toString())) { skip = true; break; }
                    if (skip) continue;
                    walk(entry, recursive, excludePatterns, sizeMap);
                } else {
                    boolean skip = false;
                    for (String pat : excludePatterns)
                        if (Pattern.matches(pat, entry.getFileName().toString())) { skip = true; break; }
                    if (skip) continue;
                    long size = Files.size(entry);
                    if (size > 0) {
                        if (!sizeMap.containsKey(size)) sizeMap.put(size, new ArrayList<>());
                        sizeMap.get(size).add(entry.toString());
                    }
                }
            }
        }
    }

    private static Map<String, List<List<String>>> findDuplicates(String root, boolean recursive,
            String algo, long partial, List<String> excludePatterns) throws Exception {
        Map<Long, List<String>> sizeMap = new HashMap<>();
        walk(Paths.get(root), recursive, excludePatterns, sizeMap);
        Map<String, List<List<String>>> duplicates = new HashMap<>();
        for (Map.Entry<Long, List<String>> entry : sizeMap.entrySet()) {
            if (entry.getValue().size() < 2) continue;
            Map<String, List<String>> hashMap = new HashMap<>();
            for (String f : entry.getValue()) {
                String h = getFileHash(f, algo, partial);
                if (!hashMap.containsKey(h)) hashMap.put(h, new ArrayList<>());
                hashMap.get(h).add(f);
            }
            for (Map.Entry<String, List<String>> hEntry : hashMap.entrySet()) {
                if (hEntry.getValue().size() > 1) {
                    if (!duplicates.containsKey(hEntry.getKey()))
                        duplicates.put(hEntry.getKey(), new ArrayList<>());
                    duplicates.get(hEntry.getKey()).add(hEntry.getValue());
                }
            }
        }
        return duplicates;
    }

    private static String formatSize(long size) {
        String[] units = {"B", "KB", "MB", "GB", "TB"};
        double s = size;
        int i = 0;
        while (s >= 1024 && i < units.length-1) { s /= 1024; i++; }
        return String.format("%.1f %s", s, units[i]);
    }

    private static void printDuplicates(Map<String, List<List<String>>> duplicates, boolean verbose) {
        if (duplicates.isEmpty()) {
            System.out.println(colorize("Дубликатов не найдено.", GREEN));
            return;
        }
        int totalGroups = 0, totalFiles = 0;
        long totalSize = 0;
        for (Map.Entry<String, List<List<String>>> entry : duplicates.entrySet()) {
            for (List<String> group : entry.getValue()) {
                totalGroups++;
                totalFiles += group.size();
                long size = 0;
                try { size = Files.size(Paths.get(group.get(0))); } catch (IOException ignored) {}
                totalSize += size * (group.size() - 1);
                System.out.println(colorize("\nГруппа дубликатов (размер: " + formatSize(size) + "):", BLUE));
                System.out.println("  Хеш: " + entry.getKey());
                for (int i = 0; i < group.size(); i++) {
                    if (i == 0)
                        System.out.println(colorize("  [оригинал] " + group.get(i), GREEN));
                    else
                        System.out.println(colorize("  [дубликат] " + group.get(i), YELLOW));
                }
                if (verbose) System.out.println("  Количество файлов: " + group.size());
            }
        }
        System.out.println(colorize("\nВсего групп: " + totalGroups + ", файлов: " + totalFiles +
                                    ", экономия: " + formatSize(totalSize), GREEN));
    }

    public static void main(String[] args) {
        String root = ".";
        boolean recursive = true, deleteDups = false, hardlink = false, dryRun = false, verbose = false, yes = false;
        String algo = "md5", moveTo = null, logFile = null;
        long partial = 0;
        List<String> excludePatterns = new ArrayList<>();

        for (int i = 0; i < args.length; i++) {
            String arg = args[i];
            if (arg.equals("-p") && i+1 < args.length) root = args[++i];
            else if (arg.equals("-r")) recursive = true;
            else if (arg.equals("-a") && i+1 < args.length) algo = args[++i];
            else if (arg.equals("--partial") && i+1 < args.length) partial = Long.parseLong(args[++i]);
            else if (arg.equals("-e") && i+1 < args.length) excludePatterns.addAll(Arrays.asList(args[++i].split(",")));
            else if (arg.equals("--delete")) deleteDups = true;
            else if (arg.equals("--move") && i+1 < args.length) moveTo = args[++i];
            else if (arg.equals("--hardlink")) hardlink = true;
            else if (arg.equals("-n")) dryRun = true;
            else if (arg.equals("-l") && i+1 < args.length) logFile = args[++i];
            else if (arg.equals("-v")) verbose = true;
            else if (arg.equals("-y")) yes = true;
            else if (arg.equals("-h")) { System.out.println("Help..."); return; }
        }

        try {
            if (!Files.isDirectory(Paths.get(root)))
                throw new Exception("'"+root+"' не является папкой");
            Map<String, List<List<String>>> duplicates = findDuplicates(root, recursive, algo, partial, excludePatterns);
            printDuplicates(duplicates, verbose);

            if (deleteDups || moveTo != null || hardlink) {
                if (!yes) {
                    System.out.print(colorize("Выполнить действия над дубликатами? [y/N] ", YELLOW));
                    Scanner sc = new Scanner(System.in);
                    String ans = sc.nextLine();
                    if (!ans.equalsIgnoreCase("y")) {
                        System.out.println(colorize("Операция отменена.", RED));
                        return;
                    }
                }
                for (List<List<String>> groups : duplicates.values()) {
                    for (List<String> group : groups) {
                        String original = group.get(0);
                        for (int i = 1; i < group.size(); i++) {
                            String dup = group.get(i);
                            if (dryRun) {
                                System.out.println(colorize("[DRY RUN] Будет обработан дубликат: " + dup, YELLOW));
                                continue;
                            }
                            if (deleteDups) {
                                Files.delete(Paths.get(dup));
                                System.out.println(colorize("Удалён: " + dup, RED));
                            } else if (moveTo != null) {
                                Files.createDirectories(Paths.get(moveTo));
                                Path dest = Paths.get(moveTo, Paths.get(dup).getFileName().toString());
                                Files.move(Paths.get(dup), dest);
                                System.out.println(colorize("Перемещён: " + dup + " -> " + dest, YELLOW));
                            } else if (hardlink) {
                                Files.delete(Paths.get(dup));
                                Files.createLink(Paths.get(dup), Paths.get(original));
                                System.out.println(colorize("Создана жёсткая ссылка: " + dup + " -> " + original, BLUE));
                            }
                        }
                    }
                }
            }
        } catch (Exception e) {
            System.err.println(colorize("Ошибка: " + e.getMessage(), RED));
        }
    }
}
